### DISCUSSION TITLE
Algorithm for branch url truncation

### DISCUSSION THREAD

**Discussion by KevinKreps** [vercel/community#472] [Help] [answered]
Is there an easy way to get the automatic-branch-url in a github action? Or where can I find the exact algorithm for the url generation including truncation?
https://vercel.com/docs/concepts/deployments/automatic-urls#truncation

**Comment by leo** [comment, marked as answer]
Thanks for reaching out about this!

At the moment, we don't provide a utility function for this, but we're looking to expose these URLs as [System Environment Variables](https://vercel.com/docs/concepts/projects/environment-variables#system-environment-variables) in the future to solve this problem for you.

In the meantime, however, truncation should be extremely rare. It only applies for really long combinations of slugs, as mentioned in the documentation, so you should be fine if you try to construct e.g. the Automatic Branch URL from the branch, project, and team name yourself.

**Reply by KevinKreps** [reply]
Thanks for your answer!
OK, that is bad as branch names are mostly very verbose in our project, but it's possible to deal with.
What about special characters like slash, space and underscore etc? Is there any information in the documentation how these are handled? If I stitch the URL together manually, then this is the remaining critical information I need. It looks like "-" is used a lot instead, and I would like to make sure all the edge cases if possible.

**Reply by AndyBitz** [reply]
For truncation, we first replace all the slashes with dashes. After that, the new string is made URL safe, [this NPM package](https://www.npmjs.com/package/url-safe-string) should achieve that for you. Finally, we trim all leading dashes and trailing dashes.

**Reply by havenchyk** [reply]
@AndyBitz what about adding hash in the end? how can I get that postfix?

**Reply by havenchyk** [reply]
@leo any ETA when those env variables will be exposed? I need to connect two different vercel deployments with random URLs

**Reply by AndyBitz** [reply]
The hashing is explained [here](https://vercel.com/docs/concepts/deployments/automatic-urls#truncation) in the docs. It's a sha256 of the branch + project name.

Leo is not with Vercel anymore, so he won't be able to answer. Maybe @simpsoka can help.

> I need to connect two different vercel deployments with random URLs

Would you mind explaining further? Are you creating two Deployments on different branches that need to be aware of each other, and do those branches change?

**Reply by havenchyk** [reply]
Thanks @AndyBitz for answering. Actually I deploy client and server separately. They both are projects on vercel. client needs to know server URL to make HTTP requests to it. since I don't control automatic preview URLs, I have to use my `client-git-branch-team.vercel.app` and just replace `client` with `server`, so server URL I predict as `server-git-branch-team.vercel.app`. Unfortunately dependabot URLs are long enough, so I come across truncation with them. And it's hard to predict how server URL will look like as it's `server-git-very-very-long-branch-na-sd3fd3-team.vercel.app`. 

Seems I can use a separate [github action](https://github.com/marketplace/actions/vercel-action) to make deployment on my own, and get server URL after deployment, then pass it as ENV variable to client.

**Reply by havenchyk** [reply]
E.g. I have a branch `dependabot-npm_and_yarn-client-types-react-avatar-editor-12.0.0`, I have my project called `client`,

and the resulted URL looks like `https://client-git-dependabot-npmandyarn-client-types-re-6354db-my-team.vercel.app`

`sha256('dependabot-npm_and_yarn-client-types-react-avatar-editor-12.0.0' + 'client')` doesn't contain `6354db`

**Reply by AndyBitz** [reply]
@havenchyk you're correct, the docs are incomplete, I'll update them right away.

We actually prefix the branch before we hash it, so you'll need to do `sha256('git-' + branch + projectName)`. Or, in case it's a fork, you'll need to use `git-fork-` instead of just `git-`.

So this should work for you:

```typescript
  const projectName = 'client';
  const branch = 'dependabot-npm_and_yarn-client-types-react-avatar-editor-12.0.0';

  const hash = require('crypto').createHash('sha256')
      .update(`git-${branch}${projectName}`)
      .digest('hex')
      .slice(0, 6);

  console.log(hash === '6354db');
```

**Reply by havenchyk** [reply]
@AndyBitz  it should help a lot, thank you 😄 

it would be nice to know the truncation algorithm

**Reply by lynndylanhurley** [reply]
@havenchyk this seems to produce the same results as vercel:

```ts
import crypto from 'crypto';

const slugify = (str: string) => {
  const slug = str
    .toLowerCase() // Convert to lowercase
    .replace(/[^a-z0-9-]/g, '-') // Replace any non-alphanumeric characters with hyphens
    .replace(/-+/g, '-') // Replace any consecutive hyphens with a single hyphen
    .replace(/^-|-$/g, '') // Remove any leading or trailing hyphens
    .replace(/\//g, '-'); // replace forward slashes with hyphens
  return slug;
};

export const truncate = (
  prefix: string,
  branch: string,
  projectName: string,
  org: string
) => {
  const branchSlug = slugify(branch);
  const deploymentName = `${projectName}-${prefix}-${branchSlug}-${org}`;
  if (deploymentName.length < 64) {
    return `https://${deploymentName}.vercel.app`;
  }

  const hash = crypto
    .createHash('sha256')
    .update(`git-${branch}${projectName}`)
    .digest('hex')
    .slice(0, 6);

  const suffix = `${hash.slice(0, 6)}-${org}`;
  const truncatedName = deploymentName.slice(0, 61 - suffix.length);
  return `https://${truncatedName}-${suffix}.vercel.app`;
};
```
